### PR TITLE
Support of en/decapsulation in the pkeyutl command

### LINKED ISSUE #25002: Provide support for using KEMs in command line utils
while it's possible to generate a private key for ML-KEM-512 or Kyber-512 (if they're included in the osqprovider) using the `openssl genpkey` utility, like so:
```
openssl genpkey -provider default -provider oqsprovider -out /tmp/kyber512.pem -algorithm kyber512
```
and it's possible to convert it then to a public key:
```
openssl pkey -provider default -provider oqsprovider -pubout -in /tmp/kyber512.pem -out /tmp/pub.pem
```
there is no way to actually perform the encapsulate operation with them...

if I try to perform a Diffie-Hellman like derive, the operation fails:
```
# openssl pkeyutl -provider oqsprovider -provider default -derive -out /tmp/shared.key -inkey /tmp/ml_kem_512.pem 
pkeyutl: Error initializing context
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:04000066:object identifier routines:OBJ_create:oid exists:crypto/objects/obj_dat.c:781:
80527EDC987F0000:error:03000096:digital envelope routines:EVP_PKEY_derive_init_ex:operation not supported for this keytype:crypto/evp/exchange.c:362:
```

in fact, if I look for the `EVP_PKEY_encapsulate()` API in `apps/` I see it only in `speed.c`...

### PR DESCRIPTION
**Pull request by beldmit**
Fixes #25002

@tomato42 if you are able to take a look it would be great

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by beldmit** [comment]
Tests added

**Comment by beldmit** [comment]
Ready for review

**Review by nhorman — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

File: doc/man1/openssl-pkeyutl.pod.in [hunk 1]
```diff
@@ -79,6 +83,10 @@ then the B<-rawin> option must be also specified.
 Specifies the output filename to write to or standard output by
 default.
 
+=item B<-secret> I<filename>
+
+Specifies the output filename to write the secret to on I<-encap>.
```
Comment (vdukhovni) [inline doc/man1/openssl-pkeyutl.pod.in:88]: Isn't this also used (optionally) for the decapsulated output???
```suggestion
Specifies the output filename to write the I<encapsulated> secret to with I<-encap> or I<decapsulated> secret with I<-decap>.
```

File: doc/man1/openssl-pkeyutl.pod.in [hunk 2]
```diff
@@ -147,6 +155,19 @@ Decrypt the input data using a private key.
 
 Derive a shared secret using the peer key.
 
+=item B<-encap>
+
+Encapsulate the secret using a public key. I<-secret> is required.
```
Comment (vdukhovni) [inline doc/man1/openssl-pkeyutl.pod.in:160]:
```suggestion
Encapsulate the secret using a public key.
The I<-secret> option must also be provided to specify the output file.
```

File: doc/man1/openssl-pkeyutl.pod.in [hunk 3]
```diff
@@ -147,6 +155,19 @@ Decrypt the input data using a private key.
...
+
+=item B<-decap>
+
+Decapsulate the secret using a private key.
```
Comment (vdukhovni) [inline doc/man1/openssl-pkeyutl.pod.in:164]:
```suggestion
Decapsulate the secret using a private key.
The result (binary data) is written to standard output by default, or else to the file specified with I<-secret>.
```

File: doc/man1/openssl-pkeyutl.pod.in [hunk 4]
```diff
@@ -147,6 +155,19 @@ Decrypt the input data using a private key.
...
+
+=item B<-kemop> I<operation>
+
+KEM operation, specific for the key algorithm. See L<EVP_PKEY_CTX_set_kem_op(3)>
+and algorithm-specific KEM documentation.
```
Comment (vdukhovni) [inline doc/man1/openssl-pkeyutl.pod.in:169]: Can anything be said here briefly to explain what this does, or when it is needed?

File: test/decap_out.bin [hunk 5]
```diff
@@ -0,0 +1,3 @@
+6�W�����n���;������m� ĥ�B[H���#�Ӈ(��h�]:\�P��x�e���b��)G�f��"��˭f�	��J�)�����{�Hm�\Pú�+�P޸%��/jϙ%�؆�<_�~�
+K�JEh����lEa�:�(��/\Ѯ�b��î����-g,AY��4�
+l��t�N�)~\�HU4y�	}qJ���t#��}.��T���?��ϊ��cD=�L��nmv�{��ſԋȣ�
```
Comment (vdukhovni) [inline test/decap_out.bin:3]: Mark this and similar files as binary blobs in git (unless that's already the case), I would not expect to see the content presented as text for binary blobs.

**Review by vdukhovni — CHANGES_REQUESTED** [review]

File: doc/man1/openssl-pkeyutl.pod.in (hunk 1 above)
Comment (beldmit) [reply doc/man1/openssl-pkeyutl.pod.in:88]: No, on `-decap` the current implementation uses only `-out` file for this purpose.

**Comment by beldmit** [comment]
@vdukhovni I've expanded the documentation according to you review, could you please take another look?

File: doc/man1/openssl-pkeyutl.pod.in [hunk 6]
```diff
@@ -147,6 +155,28 @@ Decrypt the input data using a private key.
 
 Derive a shared secret using the peer key.
 
+=item B<-encap>
+
+Encapsulate the secret using a private key. The result (binary data) is written
+to standard output by default, or else to the file specified with I<-out>. The
+I<-secret> option must also be provided to specify the secret value generated
+in the encapsulation process.
```
Comment (vdukhovni) [inline doc/man1/openssl-pkeyutl.pod.in:163]: It seems that the secret value is also *generated* as part of the encapsulation operation (i.e. not provided, but is additional output), then the description can be more clear:
```suggestion
Encapsulate a generated secret using a private key.
The encapsulated result (binary data) is written
to standard output by default, or else to the file specified with I<-out>.
The I<-secret> option must also be provided to specify the output file for the secret value generated
in the encapsulation process.
```
(Fold long lines as necessary). Per `mandoc` specs, start each sentence on a new line.

**Review by vdukhovni — CHANGES_REQUESTED** [review]

File: doc/man1/openssl-pkeyutl.pod.in (hunk 6 above)
Comment (beldmit) [reply doc/man1/openssl-pkeyutl.pod.in]: Thanks, accepted

File: apps/pkeyutl.c [hunk 7]
```diff
@@ -402,6 +420,15 @@ int pkeyutl_main(int argc, char **argv)
...
         goto end;
 
+    if (secoutfile == NULL && pkey_op == EVP_PKEY_OP_ENCAPSULATE) {
+        BIO_printf(bio_err, "Encapsulation requires '-secret' argument\n");
+        goto end;
+    } else {
+        secout = bio_open_default(secoutfile, 'w', FORMAT_BINARY);
+        if (secout == NULL)
+            goto end;
+    }
```
Comment (vdukhovni) [inline apps/pkeyutl.c:430]: Is this needed when the `pkey_op` is `EVP_PKEY_OP_DECAPSULATE`?
Perhaps the `scout` file should not be opened in that case?
The documentation says that the decapsulated secret is written to the `-out` file...

**Review by vdukhovni — CHANGES_REQUESTED** [review]
Final nit I think, should be able to approve once this is cleared up.

File: apps/pkeyutl.c (hunk 7 above)
Comment (beldmit) [reply apps/pkeyutl.c]: I don't think it's crucial but for clarity I changed the code to open this file only for encapsulation

File: apps/pkeyutl.c [hunk 8]
```diff
@@ -402,6 +420,17 @@ int pkeyutl_main(int argc, char **argv)
...
+        BIO_printf(bio_err, "Encapsulation requires '-secret' argument\n");
+        goto end;
+    } else {
+        if (pkey_op == EVP_PKEY_OP_ENCAPSULATE) {
+            secout = bio_open_default(secoutfile, 'w', FORMAT_BINARY);
+            if (secout == NULL)
+                goto end;
+        }
+    }
+
```
Comment (vdukhovni) [inline apps/pkeyutl.c:433]:
```suggestion
    if (pkey_op == EVP_PKEY_OP_ENCAPSULATE) {
        if (secoutfile == NULL) {
            BIO_printf(bio_err, "Encapsulation requires '-secret' argument\n");
            goto end;
        }
        secout = bio_open_default(secoutfile, 'w', FORMAT_BINARY);
        if (secout == NULL)
            goto end;
    }

```

**Review by vdukhovni — CHANGES_REQUESTED** [review]
Suggesting a simplification of the updated code, and then we're done.

File: apps/pkeyutl.c (hunk 8 above)
Comment (beldmit) [reply apps/pkeyutl.c]: Thanks, applied

**Review by vdukhovni — APPROVED** [review]

**Comment by beldmit** [comment]
@paulidale @nhorman please reapprove

**Comment by beldmit** [comment]
The test failure is irrelevant

**Comment by nhorman** [comment]
ack, approval holds

**Review by paulidale — APPROVED** [review]
still good

**Comment by beldmit** [comment]
Merged. Thanks for the review!

**Comment by tomato42** [comment]
sorry for late reply, was on holidays

that being said, LGTM! :)

**Pull request closed without merging**